File: src/main/java/com/github/solairerove/algs4/leprosorium/arrays/SmallestDifference.java
package com.github.solairerove.algs4.leprosorium.arrays;

import java.util.Arrays;

/**
 * Takes two non-empty arrays, finds pair with difference is closest to zero.
 */
public class SmallestDifference {
    public static void main(String... args) {
        int[] a = {1, 5, 3, 19, 18, 25};
        int[] b = {12, 20, -4, 31, 26, 100};

        System.out.println(Arrays.toString(smallestDifference(a, b)));
    }

    // O(nlog(n) + mlog(m)) time | O(1) space
    private static int[] smallestDifference(int[] a, int[] b) {
        Arrays.sort(a);
        Arrays.sort(b);

        int idxOne = 0, idxTwo = 0;
        int smallest = Integer.MAX_VALUE;
        int[] smallestPair = new int[2];

        while (idxOne < a.length && idxTwo < b.length) {
            int firstNum = a[idxOne], secondNum = b[idxTwo];
            var current = Math.abs(secondNum - firstNum);

            if (firstNum < secondNum) idxOne++;
            else if (firstNum > secondNum) idxTwo++;
            else return new int[]{firstNum, secondNum};

            if (current < smallest) {
                smallest = current;
                smallestPair = new int[]{firstNum, secondNum};
            }
        }

        return smallestPair;
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/arrays/SmallestDifference.kt
package com.github.solairerove.algs4.leprosorium.arrays

fun main() {
    val arrOne = mutableListOf(2, -9, 0, 14, 1, -8, -3, -1, 8, -6, 5, -5)
    val arrTwo = mutableListOf(15, -8, 3, 1, 7, -10, 2, -1, 4, -3, 17, -2)

    print(smallestDifference(arrOne, arrTwo)) // [-8, -8]
}

// O(nlog(n) + mlog(m)) time | O(1) space
private fun smallestDifference(arrOne: MutableList<Int>, arrTwo: MutableList<Int>): List<Int> {
    arrOne.sort()
    arrTwo.sort()

    var smallestPair = listOf<Int>()
    var idxOne = 0
    var idxTwo = 0
    var smallest = Int.MAX_VALUE
    var curr: Int

    while (idxOne < arrOne.size && idxTwo < arrTwo.size) {
        val numOne = arrOne[idxOne]
        val numTwo = arrTwo[idxTwo]

        when {
            numOne < numTwo -> {
                curr = numTwo - numOne
                idxOne++
            }

            numTwo < numOne -> {
                curr = numOne - numTwo
                idxTwo++
            }

            else -> return listOf(numOne, numTwo)
        }

        if (curr < smallest) {
            smallest = curr
            smallestPair = listOf(numOne, numTwo)
        }
    }

    return smallestPair
}
